src/systems/buoyancy_engine/BuoyancyEngine.cc

```diff
@@ -0,0 +1,224 @@
+  this->volumeSetPoint = volume;
+  igndbg << "Updating volume " << volume <<"\n";
+}
+
+//////////////////////////////////////////////////
+BuoyancyEnginePlugin::BuoyancyEnginePlugin()
+  : dataPtr(std::make_unique<BuoyancyEnginePrivateData>())
+{
+}
+
+//////////////////////////////////////////////////
+void BuoyancyEnginePlugin::Configure(
+  const ignition::gazebo::Entity &_entity,
+  const std::shared_ptr<const sdf::Element> &_sdf,
+  ignition::gazebo::EntityComponentManager &_ecm,
+  ignition::gazebo::EventManager &/*_eventMgr*/)
+{
+  auto model = ignition::gazebo::Model(_entity);
+  if (!_sdf->HasElement("link_name"))
+  {
+    ignerr << "Buoyancy Engine must be attached to some link."  << std::endl;
+    return;
+  }
+  this->dataPtr->linkEntity =
+    model.LinkByName(_ecm, _sdf->Get<std::string>("link_name"));
+
+  if (_sdf->HasElement("min_volume"))
+  {
+    this->dataPtr->minVolume = _sdf->Get<double>("min_volume");
+  }
+
+  if (_sdf->HasElement("max_volume"))
+  {
+    this->dataPtr->maxVolume = _sdf->Get<double>("max_volume");
+  }
+
+  if (_sdf->HasElement("fluid_density"))
+  {
+    this->dataPtr->fluidDensity = _sdf->Get<double>("fluid_density");
+  }
+
+  this->dataPtr->bladderVolume = this->dataPtr->minVolume;
+  if (_sdf->HasElement("default_volume"))
+  {
+    this->dataPtr->bladderVolume = _sdf->Get<double>("default_volume");
+    this->dataPtr->volumeSetPoint = this->dataPtr->bladderVolume;
+  }
+
+  if (_sdf->HasElement("neutral_volume"))
+  {
+    this->dataPtr->neutralVolume = _sdf->Get<double>("neutral_volume");
+  }
+
+  std::string cmdTopic = "/buoyancy_engine/";
+  std::string statusTopic = "/buoyancy_engine/current_volume";
+  if (_sdf->HasElement("namespace"))
+  {
+    cmdTopic = ignition::transport::TopicUtils::AsValidTopic(
+      "/model/" + _sdf->Get<std::string>("namespace") + "/buoyancy_engine/");
+    statusTopic = ignition::transport::TopicUtils::AsValidTopic(
+      "/model/" + _sdf->Get<std::string>("namespace")
+      + "/buoyancy_engine/current_volume");
+  }
+  igndbg << "listening on topic: " << cmdTopic <<std::endl;
+
+  if(_sdf->HasElement("max_inflation_rate"))
+  {
+    this->dataPtr->maxInflationRate = _sdf->Get<double>("max_inflation_rate");
+  }
+
+  if(!this->dataPtr->node.Subscribe(cmdTopic,
+    &BuoyancyEnginePrivateData::OnCmdBuoyancyEngine, this->dataPtr.get()))
+  {
+    ignerr << "Failed to subscribe to [" << cmdTopic << "]" << std::endl;
+  }
+
+  this->dataPtr->statusPub =
+    this->dataPtr->node.Advertise<ignition::msgs::Double>(statusTopic);
+}
+
+//////////////////////////////////////////////////
+void BuoyancyEnginePlugin::PreUpdate(
+  const ignition::gazebo::UpdateInfo &_info,
+  ignition::gazebo::EntityComponentManager &_ecm)
+{
+  if (_info.paused)
+    return;
+
+  typedef std::chrono::duration<float, std::ratio<1L, 1L>> DurationInSecs;
+  auto dt = std::chrono::duration_cast<DurationInSecs>(_info.dt).count();
+
+  ignition::msgs::Double msg;
+
+  double forceMag;
```

Comment:
Just a nitpicky comment about the variable name. I associated "mag" to magnitude and assumed this value would be always positive, which is not the case. Maybe `zForce` or something would be clearer.

Reply:
Resolved in 8ba2e0e